File: source_code/python/bfsgraph.py
from collections import deque
from node import node

def bfsgraph(s0):
    # s0 = initial state

    # create the initial node
    n0 = node(s0, None, None, 0, 0) 

    # initizlize #visited nodes
    nvisited = 0
    
    # initialize the frontier list
    frontier = deque([n0])

    # initialize the explored set
    explored = set()

    while True:
        # the search fails when the frontier is empty
        if not frontier:
            return (None, nvisited)
        else:
            # get one node from the frontier
            n = frontier.popleft()
            # add the state to the expolored set
            explored.add(str(n.state))
            # count the number of visited nodes
            nvisited+=1
            # check if the state in n is a goal
            if n.state.isgoal():
                return (n, nvisited)
            else:
                # generate successor states
                S = n.state.successors()
                # create new nodes and add to the frontier
                for (s, a, c) in S:
                    if str(s) not in explored and not find_state(s, frontier): 
                        p = node(s, n, a, n.cost+c, n.depth+1)
                        frontier.append(p)

def find_state(s, l):
    for n in l:
        if str(n.state) == str(s):
            return True
    return False

File: source_code/java/BFSGraph.java
import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;

public class BFSGraph {
  public int nvisited;

  public Node search(State s0) {
    Node n0 = new Node(s0, null, null, 0, 0);
    nvisited = 0;
    ArrayList<Node> frontier = new ArrayList<Node>();
    HashSet<String> explored = new HashSet<String>();
    frontier.add(n0);
    while(true) {
      if (frontier.isEmpty()) {
        return null;
      }
      else {
        Node n = frontier.remove(0);
        explored.add(n.state.toString());
        nvisited++;
        if (n.state.isGoal()) {
          return n;
        }
        else {
          List<Successor> succs = n.state.successors();
          for(Successor s : succs) {
            if (!explored.contains(s.state.toString()) &&
                !containsState(s.state, frontier)) { 
              Node p = new Node(s.state, n, s.action, n.cost+s.cost, n.depth+1);
              frontier.add(p);
            }
          }
        }
      }
    }
  }

  public boolean containsState(State s, List<Node> l) {
    for(Node n : l) {
      if (n.state.toString().equals(s.toString())) {
        return true;
      }
    }
    return false;
  }
}
